test/Npgsql.Tests/PrepareTests.cs

```diff
@@ -608,6 +609,63 @@ public void SameSqlDifferentParams()
         }
         */
 
+        [Test]
+        public void InvalidStatement()
+        {
+            using (var conn = OpenConnection())
+            {
+                var cmd = new NpgsqlCommand("sele", conn);
+                Assert.That(() => cmd.Prepare(), Throws.Exception.TypeOf<PostgresException>());
+            }
+        }
+
+        [Test]
+        public void PrepareMultipleCommandsWithParameters()
+        {
+            using (var conn = OpenConnection())
+            {
+                using (var cmd1 = new NpgsqlCommand("SELECT @p1;", conn))
+                using (var cmd2 = new NpgsqlCommand("SELECT @p1; SELECT @p2;", conn))
+                {
+                    var p1 = new NpgsqlParameter("p1", NpgsqlDbType.Integer);
+                    var p21 = new NpgsqlParameter("p1", NpgsqlDbType.Text);
+                    var p22 = new NpgsqlParameter("p2", NpgsqlDbType.Text);
+                    cmd1.Parameters.Add(p1);
+                    cmd2.Parameters.Add(p21);
+                    cmd2.Parameters.Add(p22);
+                    cmd1.Prepare();
+                    cmd2.Prepare();
+                    p1.Value = 8;
+                    p21.Value = "foo";
+                    p22.Value = "bar";
+                    using (var reader1 = cmd1.ExecuteReader())
+                    {
+                        Assert.That(reader1.Read(), Is.True);
+                        Assert.That(reader1.GetInt32(0), Is.EqualTo(8));
+                    }
+                    using (var reader2 = cmd2.ExecuteReader())
+                    {
+                        Assert.That(reader2.Read(), Is.True);
+                        Assert.That(reader2.GetString(0), Is.EqualTo("foo"));
+                        Assert.That(reader2.NextResult(), Is.True);
+                        Assert.That(reader2.Read(), Is.True);
+                        Assert.That(reader2.GetString(0), Is.EqualTo("bar"));
+                    }
+                }
+            }
+        }
+
+        [Test]
+        public void MultiplexingNotSupported()
+        {
+            var builder = new NpgsqlConnectionStringBuilder(ConnectionString) { Multiplexing = true };
+            using var conn = OpenConnection(builder);
+            using var cmd = new NpgsqlCommand("SELECT 1", conn);
+
+            Assert.That(() => cmd.Prepare(), Throws.Exception.TypeOf<NotSupportedException>());
```

Comment:
Explicit is unsupported - but auto-prepare is completely supported.

I do have an item in my scoped-out TODO for explicit preparation. I see two options here:
* Simply bind a connector when prepare is called, and then do normal explicit prepare. That makes things work more or less like today
* More radical: mark the command for preparation, thereby bypassing auto-prepare on the other side and preparing right away. That means we don't need to get to the min threshold, and the SQL never gets unprepared because of LRU.

Let me know what you think.

Reply:
I would personally like to see the latter (I see quite a few uses for explicit prepare and multiplexing) Though there are cases where you just want to prepare on a dedicated connection, complicating the preparation api to accommodate both wouldn't be great, though maybe people can just bind to a connector first?.

Can you explain what you mean by (I haven't done a lot of reading of the auto preparation code)
>and the SQL never gets unprepared because of LRU

Reply:
> Can you explain what you mean by

Auto-preparation is basically an LRU cache - Max Auto Prepare defines the maximum number of (auto-prepared) prepared statements we have at any given point. When we decide we want to auto-prepare a new SQL statement (because it's been executed a lot lately), we evict the LRU one and replace it. So the explicit preparation API (NpgsqlCommand.Prepare) could be redefined to bypass this mechanism - prepare the statement immediately and never evict it, which would be similar to how explicit prepared statements work today without multplexing. The auto-prepared list is completely separate from explicitly-prepared statements - in addition to Max AutoPrepare you can have as many explicitly prepared statements as you want (since you're managing them explicitly anyway).

Reply:
and the SQL never gets unprepared because of LRU

Then that was the confusion I had, I thought LRU somehow prevented depreparation 100% of the time, which didn't make sense to me considering what a LRU cache is.